pkg/api/vminfo.go

```diff
@@ -129,3 +151,63 @@ func ExtractVmInfo(vm *vmapi.VirtualMachine) (*VmInfo, error) {
 
 	return &info, nil
 }
+
+func (vm VmInfo) EqualScalingBounds(cmp VmInfo) bool {
+	return vm.Min() != cmp.Min() || vm.Max() != cmp.Max()
+}
+
+type scalingBounds struct {
+	Min *resourceBound `json:"min,omitempty"`
+	Max *resourceBound `json:"max,omitempty"`
+}
+
+type resourceBound struct {
+	CPU *uint16            `json:"cpu,omitempty"`
+	Mem *resource.Quantity `json:"mem,omitempty"`
+}
+
+func (b scalingBounds) validate(memSlotSize *resource.Quantity) error {
+	if b.Min == nil {
+		return errors.New("missing field 'min'")
+	} else if b.Max == nil {
+		return errors.New("missing field 'max'")
+	}
+
+	if field, err := b.Min.validate(memSlotSize); err != nil {
+		return fmt.Errorf("error at .min%s: %w", field, err)
+	} else if field, err := b.Max.validate(memSlotSize); err != nil {
+		return fmt.Errorf("error at .max%s: %w", field, err)
+	}
+
+	return nil
+}
+
+func (b resourceBound) validate(memSlotSize *resource.Quantity) (field string, _ error) {
+	if b.CPU == nil {
+		return "", errors.New("missing field 'cpu'")
+	} else if b.Mem == nil {
+		return "", errors.New("missing field 'mem'")
+	}
+
+	if *b.CPU == 0 {
+		return ".cpu", errors.New("value cannot be zero")
+	}
+
+	if b.Mem.IsZero() || b.Mem.Value() < 0 {
+		return ".mem", errors.New("value must be greater than zero")
+	} else if b.Mem.Value()%memSlotSize.Value() != 0 {
+		return ".mem", fmt.Errorf("value must be divisible by VM memory slot size %s", memSlotSize)
+	}
+
+	return "", nil
+}
+
+func (b scalingBounds) apply(info *VmInfo) {
```

Comment:
might it make sense to have this be a method on `VmInfo` rather than the other way around